enterprise-metrics/main.libsonnet

```diff
@@ -0,0 +1,379 @@
+  tokengen: {
+    local target = 'tokengen',
+    '#args':: d.obj('`args` is convenience field for modifying the tokegen container arguments as key-value pairs.'),
+    args:: this._config.commonArgs {
+      target: target,
+      'tokengen.token-file': '/shared/admin-token',
+    },
+    '#container':: d.obj(|||
+      `container` is a convenience field for modifying the tokengen container.
+      By default, the container runs GEM with the tokengen target and writes the token to a file.
+    |||),
+    container::
+      container.new(target, this._images.gem)
+      + container.withPorts(cortex.util.defaultPorts)
+      + container.withArgs(util.mapToFlags(self.args))
+      + container.withVolumeMounts([{ mountPath: '/shared', name: 'shared' }])
+      + container.resources.withLimits({ memory: '4Gi' })
+      + container.resources.withRequests({ cpu: '500m', memory: '500Mi' }),
+
+    '#createSecretContainer':: d.obj('`createSecretContainer` creates a Kubernetes Secret from a token file.'),
+    createSecretContainer::
+      container.new('create-secret', this._images.kubectl)
+      + container.withCommand([
+        '/bin/bash',
+        '-euc',
+        'kubectl create secret generic gem-admin-token -from-file=token=/shared/admin-token -from-literal=grafana-token="$(base64 <(echo :$(cat /shared/admin-token)))"',
+      ])
+      + container.withVolumeMounts([{ mountPath: '/shared', name: 'shared' }])
+      // Need to run as root because the GEM container does.
+      + container.securityContext.withRunAsUser(0),
+
+    '#job':: d.obj('`job` is the Kubernetes Job for tokengen'),
+    job:
+      job.new(target)
+      + job.spec.template.spec.withContainers([self.createSecretContainer])
+      + job.spec.template.spec.withInitContainers([self.container])
+      + job.spec.template.spec.withRestartPolicy('Never')
+      + job.spec.template.spec.withServiceAccount(target)
+      + job.spec.template.spec.withServiceAccountName(target)
+      + job.spec.template.spec.withVolumes([{ name: 'shared', emptyDir: {} }]),
+
+    '#serviceAccount':: d.obj('`serviceAccount` is the Kubernetes ServiceAccount for tokengen'),
+    serviceAccount:
+      serviceAccount.new(target),
+
+    '#role':: d.obj('`role` is the Kubernetes Role for tokengen'),
+    role:
+      role.new(target)
+      + role.withRules([
+        policyRule.withApiGroups([''])
+        + policyRule.withResources(['secrets'])
+        + policyRule.withVerbs(['create']),
+      ]),
+
+    '#roleBinding':: d.obj('`roleBinding` is the Kubernetes RoleBinding for tokengen'),
+    roleBinding:
+      roleBinding.new()
+      + roleBinding.metadata.withName(target)
+      + roleBinding.roleRef.withApiGroup('rbac.authorization.k8s.io')
+      + roleBinding.roleRef.withKind('Role')
+      + roleBinding.roleRef.withName(target)
+      + roleBinding.withSubjects([
+        subject.new()
+        + subject.withName(target)
+        + subject.withKind('ServiceAccount'),
+      ]),
+  },
+}
```

Comment:
this file is large, would it make sense to split it for each tool?

Reply:
It certainly is. I'd like to work on splitting this out into smaller files but I am of the opinion that each Jsonnet file should work as a unit to help users understand the library while tooling is still relatively weak. I'm not sure it's good for files to be reaching for configuration outside of its own file which makes sharing common configuration awkward across multiple files.